[Question] Using yavijava offline (online DTD parsing)

Hello.

While trying yavijava in Internet-less environment I'm experienced exception:

```
java.rmi.RemoteException: An error occurred parsing XML with return type: ServiceContent; nested exception is: 
	java.net.ConnectException: Connection timed out: connect
	at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:89) ~[yavijava-6.0.05.jar:na]
	at com.vmware.vim25.ws.SoapClient.unMarshall(SoapClient.java:253) ~[yavijava-6.0.05.jar:na]
	at com.vmware.vim25.ws.WSClient.invoke(WSClient.java:96) ~[yavijava-6.0.05.jar:na]
	at com.vmware.vim25.ws.VimStub.retrieveServiceContent(VimStub.java:1675) [yavijava-6.0.05.jar:na]
	at com.vmware.vim25.mo.ServiceInstance.retrieveServiceContent(ServiceInstance.java:246) [yavijava-6.0.05.jar:na]
	at com.vmware.vim25.mo.ServiceInstance.constructServiceInstance(ServiceInstance.java:126) [yavijava-6.0.05.jar:na]
	at com.vmware.vim25.mo.ServiceInstance.<init>(ServiceInstance.java:79) [yavijava-6.0.05.jar:na]
	at com.vmware.vim25.mo.ServiceInstance.<init>(ServiceInstance.java:69) [yavijava-6.0.05.jar:na]
	at ru.consyst_oc.vinventory.tasks.ServerTask.run(ServerTask.java:64) [classes/:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_112]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_112]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_112]
Caused by: java.net.ConnectException: Connection timed out: connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method) ~[na:1.8.0_112]
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79) ~[na:1.8.0_112]
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[na:1.8.0_112]
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[na:1.8.0_112]
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[na:1.8.0_112]
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172) ~[na:1.8.0_112]
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[na:1.8.0_112]
	at java.net.Socket.connect(Socket.java:589) ~[na:1.8.0_112]
	at java.net.Socket.connect(Socket.java:538) ~[na:1.8.0_112]
	at sun.net.NetworkClient.doConnect(NetworkClient.java:180) ~[na:1.8.0_112]
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:432) ~[na:1.8.0_112]
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:527) ~[na:1.8.0_112]
	at sun.net.www.http.HttpClient.<init>(HttpClient.java:211) ~[na:1.8.0_112]
	at sun.net.www.http.HttpClient.New(HttpClient.java:308) ~[na:1.8.0_112]
	at sun.net.www.http.HttpClient.New(HttpClient.java:326) ~[na:1.8.0_112]
	at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1202) ~[na:1.8.0_112]
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1138) ~[na:1.8.0_112]
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1032) ~[na:1.8.0_112]
	at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:966) ~[na:1.8.0_112]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1546) ~[na:1.8.0_112]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1474) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.setupCurrentEntity(XMLEntityManager.java:647) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startEntity(XMLEntityManager.java:1304) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.impl.XMLEntityManager.startDTDEntity(XMLEntityManager.java:1270) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.setInputSource(XMLDTDScannerImpl.java:264) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.dispatch(XMLDocumentScannerImpl.java:1161) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.next(XMLDocumentScannerImpl.java:1045) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:959) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:602) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:505) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:841) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:770) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1213) ~[na:1.8.0_112]
	at com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:643) ~[na:1.8.0_112]
	at org.dom4j.io.SAXReader.read(SAXReader.java:465) ~[dom4j-1.6.1.jar:1.6.1]
	at org.dom4j.io.SAXReader.read(SAXReader.java:343) ~[dom4j-1.6.1.jar:1.6.1]
	at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:81) ~[yavijava-6.0.05.jar:na]
	... 11 common frames omitted
```

dom4j trying to validate online DTD and failed to connect to external server.
What is the proper way to disable xml validation or online DTD processing?

--- a/src/main/java/com/vmware/vim25/mo/Task.java
+++ b/src/main/java/com/vmware/vim25/mo/Task.java
@@ -196,7 +196,7 @@ public class Task extends ExtensibleManagedObject {
             if (tState.equals(TaskInfoState.running)) {
                 Thread.sleep(runningDelayInMillSecond);
             }
-            else {
+            else if (tState.equals(TaskInfoState.queued)) {
                 Thread.sleep(queuedDelayInMillSecond);
             }
         }

--- a/src/main/java/com/vmware/vim25/ws/VimStub.java
+++ b/src/main/java/com/vmware/vim25/ws/VimStub.java
@@ -580,7 +580,12 @@ public class VimStub {
         Argument[] paras = new Argument[2];
         paras[0] = new Argument("_this", "ManagedObjectReference", _this);
         paras[1] = new Argument("criteria", "DistributedVirtualSwitchPortCriteria", criteria);
-        return (String[]) getWsc().invoke("FetchDVPortKeys", paras, "String[]");
+        Object result = getWsc().invoke("FetchDVPortKeys", paras, "String[]");
+        if (result instanceof java.util.List) {
+            java.util.List<?> list = (java.util.List<?>) result;
+            return list.toArray(new String[0]);
+        }
+        return (String[]) result;
     }
 
     public DistributedVirtualPort[] fetchDVPorts(ManagedObjectReference _this, DistributedVirtualSwitchPortCriteria criteria) throws java.rmi.RemoteException, RuntimeFault {

--- a/src/main/java/com/vmware/vim25/ws/XmlGenDom.java
+++ b/src/main/java/com/vmware/vim25/ws/XmlGenDom.java
@@ -41,6 +41,8 @@ import org.dom4j.io.SAXReader;
 import org.doublecloud.ws.util.ReflectUtil;
 import org.doublecloud.ws.util.TypeUtil;
 
+import org.xml.sax.SAXException;
+
 import java.io.IOException;
 import java.io.InputStream;
 import java.lang.reflect.Array;
@@ -78,6 +80,13 @@ class XmlGenDom extends XmlGen {
         Element root = null;
         try {
             SAXReader reader = new SAXReader();
+            try {
+                reader.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+                reader.setFeature("http://xml.org/sax/features/external-general-entities", false);
+                reader.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
+            } catch (SAXException e) {
+                throw new RemoteException("Failed to configure XML parser for XXE protection", e);
+            }
             Document doc = reader.read(is);
             if(log.isTraceEnabled()) {
                 log.trace("XML Document: " + doc.asXML());

--- a/src/test/java/com/vmware/vim25/mo/TaskTest.java
+++ b/src/test/java/com/vmware/vim25/mo/TaskTest.java
@@ -1,0 +1,54 @@
+package com.vmware.vim25.mo;
+
+import com.vmware.vim25.TaskInfo;
+import com.vmware.vim25.TaskInfoState;
+import org.junit.Test;
+
+import java.rmi.RemoteException;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class TaskTest {
+
+    private static Task taskReturning(TaskInfoState state) {
+        TaskInfo info = new TaskInfo();
+        info.setState(state);
+        return new Task(null, null) {
+            @Override
+            public TaskInfo getTaskInfo() throws RemoteException {
+                return info;
+            }
+        };
+    }
+
+    @Test
+    public void waitForTask_returnsSuccessString_whenTaskSucceeds() throws Exception {
+        Task task = taskReturning(TaskInfoState.success);
+        assertEquals("success", task.waitForTask(0, 0));
+    }
+
+    @Test
+    public void waitForTask_returnsErrorString_whenTaskFails() throws Exception {
+        Task task = taskReturning(TaskInfoState.error);
+        assertEquals("error", task.waitForTask(0, 0));
+    }
+
+    @Test
+    public void waitForTask_doesNotSleepAfterTaskSucceeds() throws Exception {
+        Task task = taskReturning(TaskInfoState.success);
+        long start = System.currentTimeMillis();
+        task.waitForTask(5000, 5000);
+        long elapsed = System.currentTimeMillis() - start;
+        assertTrue("waitForTask slept after task completed; elapsed=" + elapsed + "ms", elapsed < 1000);
+    }
+
+    @Test
+    public void waitForTask_doesNotSleepAfterTaskFails() throws Exception {
+        Task task = taskReturning(TaskInfoState.error);
+        long start = System.currentTimeMillis();
+        task.waitForTask(5000, 5000);
+        long elapsed = System.currentTimeMillis() - start;
+        assertTrue("waitForTask slept after task failed; elapsed=" + elapsed + "ms", elapsed < 1000);
+    }
+}

--- a/src/test/java/com/vmware/vim25/ws/VimStubFetchDVPortKeysTest.java
+++ b/src/test/java/com/vmware/vim25/ws/VimStubFetchDVPortKeysTest.java
@@ -1,0 +1,61 @@
+package com.vmware.vim25.ws;
+
+import com.vmware.vim25.ManagedObjectReference;
+import org.junit.Test;
+
+import javax.net.ssl.TrustManager;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.rmi.RemoteException;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import static org.junit.Assert.assertArrayEquals;
+
+public class VimStubFetchDVPortKeysTest {
+
+    private static Client clientReturning(Object result) {
+        return new Client() {
+            public Object invoke(String m, Argument[] p, String r) { return result; }
+            public StringBuffer invokeAsString(String m, Argument[] p) { return null; }
+            public URL getBaseUrl() { return null; }
+            public void setBaseUrl(URL u) {}
+            public String getCookie() { return null; }
+            public void setCookie(String c) {}
+            public String getVimNameSpace() { return null; }
+            public void setVimNameSpace(String v) {}
+            public int getConnectTimeout() { return 0; }
+            public void setConnectTimeout(int t) {}
+            public int getReadTimeout() { return 0; }
+            public void setReadTimeout(int t) {}
+            public TrustManager getTrustManager() { return null; }
+            public void setSoapActionOnApiVersion(String v) {}
+            public String marshall(String m, Argument[] p) { return null; }
+            public Object unMarshall(String r, InputStream is) { return null; }
+            public StringBuffer readStream(InputStream s) throws IOException { return null; }
+        };
+    }
+
+    @Test
+    public void fetchDVPortKeys_whenServerReturnsStringArray_returnsStringArray() throws Exception {
+        VimStub stub = new VimStub(clientReturning(new String[]{"port-1", "port-2"}));
+        String[] result = stub.fetchDVPortKeys(new ManagedObjectReference(), null);
+        assertArrayEquals(new String[]{"port-1", "port-2"}, result);
+    }
+
+    @Test
+    public void fetchDVPortKeys_whenServerReturnsArrayList_returnsStringArray() throws Exception {
+        ArrayList<String> list = new ArrayList<>(Arrays.asList("port-1", "port-2"));
+        VimStub stub = new VimStub(clientReturning(list));
+        String[] result = stub.fetchDVPortKeys(new ManagedObjectReference(), null);
+        assertArrayEquals(new String[]{"port-1", "port-2"}, result);
+    }
+
+    @Test
+    public void fetchDVPortKeys_whenServerReturnsEmptyArrayList_returnsEmptyStringArray() throws Exception {
+        VimStub stub = new VimStub(clientReturning(new ArrayList<String>()));
+        String[] result = stub.fetchDVPortKeys(new ManagedObjectReference(), null);
+        assertArrayEquals(new String[0], result);
+    }
+}

--- a/src/test/java/com/vmware/vim25/ws/XmlGenDomTest.java
+++ b/src/test/java/com/vmware/vim25/ws/XmlGenDomTest.java
@@ -4,9 +4,11 @@ import com.vmware.vim25.*;
 import org.junit.Assert;
 import org.junit.Test;
 
+import java.io.ByteArrayInputStream;
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
 import java.rmi.RemoteException;
 import java.util.Objects;
 
@@ -138,5 +140,30 @@ public class XmlGenDomTest {
         InputStream inputStream = new FileInputStream(new File("src/test/resources/xml/UpdateSetWithBase64Binary.xml"));
         XmlGenDom xmlGenDom = new XmlGenDom();
         UpdateSet updateSet = (UpdateSet) xmlGenDom.fromXML("UpdateSet", inputStream);
+    }
+
+    @Test(expected = RemoteException.class)
+    public void fromXML_withDoctypeDeclaration_throwsRemoteException() throws Exception {
+        // DOCTYPE declarations must be rejected to prevent XXE attacks and offline
+        // DTD-fetch failures (issues #293, #240).
+        String payload = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
+            "<!DOCTYPE foo>\n" +
+            "<soapenv:Envelope xmlns:soapenv=\"http://schemas.xmlsoap.org/soap/envelope/\">\n" +
+            "  <soapenv:Body><foo/></soapenv:Body>\n" +
+            "</soapenv:Envelope>";
+        InputStream is = new ByteArrayInputStream(payload.getBytes(StandardCharsets.UTF_8));
+        new XmlGenDom().fromXML("String", is);
+    }
+
+    @Test(expected = RemoteException.class)
+    public void fromXML_withExternalEntityReference_throwsRemoteException() throws Exception {
+        // External entity injection must be blocked (issue #293).
+        String payload = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
+            "<!DOCTYPE foo [<!ENTITY xxe SYSTEM \"file:///etc/passwd\">]>\n" +
+            "<soapenv:Envelope xmlns:soapenv=\"http://schemas.xmlsoap.org/soap/envelope/\">\n" +
+            "  <soapenv:Body><foo>&xxe;</foo></soapenv:Body>\n" +
+            "</soapenv:Envelope>";
+        InputStream is = new ByteArrayInputStream(payload.getBytes(StandardCharsets.UTF_8));
+        new XmlGenDom().fromXML("String", is);
     }
 }

